server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletGoalState.java

```diff
@@ -0,0 +1,178 @@
+      TabletBalancer balancer, TabletManagementParameters params) {
+    Preconditions.checkArgument(Ample.DataLevel.of(tm.getTableId()) == params.getLevel(),
+        "Tablet %s not in expected level %s", tm.getExtent(), params.getLevel());
+
+    // Always follow through with assignments
+    if (currentState == TabletState.ASSIGNED) {
+      return HOSTED;
+    }
+
+    KeyExtent extent = tm.getExtent();
+    // Shutting down?
+    TabletGoalState systemGoalState = getSystemGoalState(tm, params);
+
+    if (systemGoalState == TabletGoalState.HOSTED) {
+      if (!params.isParentLevelUpgraded()) {
+        // The place where this tablet stores its metadata was not upgraded, so do not assign this
+        // tablet yet.
+        return UNASSIGNED;
+      }
+
+      if (tm.getOperationId() != null) {
+        return UNASSIGNED;
+      }
+
+      if (!params.isTableOnline(tm.getTableId())) {
+        return UNASSIGNED;
+      }
+
+      switch (tm.getHostingGoal()) {
+        case NEVER:
+          return UNASSIGNED;
+        case ONDEMAND:
+          if (!tm.getHostingRequested()) {
+            return UNASSIGNED;
+          }
+      }
+
+      TServerInstance dest = params.getMigrations().get(extent);
+      if (dest != null && tm.hasCurrent() && !dest.equals(tm.getLocation().getServerInstance())) {
+        return UNASSIGNED;
+      }
+
+      if (currentState == TabletState.HOSTED && balancer != null) {
+        // see if the balancer thinks this tablet needs to be unassigned
+
+        Preconditions.checkArgument(
+            tm.getLocation().getType() == TabletMetadata.LocationType.CURRENT,
+            "Expected current tablet location %s %s", tm.getExtent(), tm.getLocation());
+        var tsii = new TabletServerIdImpl(tm.getLocation().getServerInstance());
+
+        var resourceGroup = params.getResourceGroup(tm.getLocation().getServerInstance());
+
+        if (resourceGroup != null) {
+          var reassign = balancer.needsReassignment(new TabletBalancer.CurrentAssignment() {
+            @Override
+            public TabletId getTablet() {
+              return new TabletIdImpl(tm.getExtent());
+            }
+
+            @Override
+            public TabletServerId getTabletServer() {
+              return tsii;
+            }
+
+            @Override
+            public String getResourceGroup() {
+              return resourceGroup;
+            }
+          });
+
+          if (reassign) {
+            return UNASSIGNED;
+          }
+        } else {
+          // A tablet server should always have a resource group, however there is a race
+          // conditions where the resource group map was read before a tablet server came into
+          // existence. Another possible cause for an absent resource group is a bug in accumulo.
+          // In either case do not call the balancer for now with the assumption that the resource
+          // group will be available later. Log a message in case it is a bug.
```

Comment:
Should we just assume the default resource group in this case? In subsequent iterations the Tablet would get moved if/when the resource group is known and it's not the default.

Reply:
With the changes in the PR, the data acquisition for TGW is all corralled into one place.  Thinking if we do anything, it should be done there.  Then the rest of the code can assume things are consistent.  Also, we have made changes to remove some of the race conditions that caused this, but maybe not all.  I think for now we should try to track down and understand all race conditions that cause this so we can distinguish bugs from unavoidable race conditions.  Thinking for now that this log level should be warn and we should try to triage the cause whenever it triggers and place any mitigations in the data acquisition code in TGW.

Reply:
adlusted log level and comment in 5f7a3dd.  Eventually when we got to more focused testing of the elasticity branch, we can triage errors and warns seen in the logs during testing.  If we see this we can look into it.